# 01. Import Libraries

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

# 02. Import Data

In [2]:
# 2a) Import order data
df_ords=pd.read_csv(r'C:\Users\ponni\Achievement 4\02 Data\Original data\orders.csv',index_col=False)
# 2b) Import product data
df_prods=pd.read_csv(r'C:\Users\ponni\Achievement 4\02 Data\Original data\products.csv',index_col=False)
# 2c) Import department data
df_dep=pd.read_csv(r'C:\Users\ponni\Achievement 4\02 Data\Original data\departments.csv',index_col=False)

# 03. Data Wrangling

In [3]:
# 01.dropping eval_set column from orders.csv
df_ords_2= df_ords.drop(columns=['eval_set'])

In [4]:
df_ords_2.head(15)

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0    2539329        1             1          2                  8   
1    2398795        1             2          3                  7   
2     473747        1             3          3                 12   
3    2254736        1             4          4                  7   
4     431534        1             5          4                 15   
5    3367565        1             6          2                  7   
6     550135        1             7          1                  9   
7    3108588        1             8          1                 14   
8    2295261        1             9          1                 16   
9    2550362        1            10          4                  8   
10   1187899        1            11          4                  8   
11   2168274        2             1          2                 11   
12   1501582        2             2          5                 10   
13   1901567        2             3          1                 10   
14    738281        2             4          2                 10   

    days_since_prior_order  
0                      NaN  
1                     15.0  
2                     21.0  
3                     29.0  
4                     28.0  
5                     19.0  
6                     20.0  
7                     14.0  
8                      0.0  
9                     30.0  
10                    14.0  
11                     NaN  
12                    10.0  
13                     3.0  
14                     8.0

In [5]:
# Investigating missing values
df_ords_2['days_since_prior_order'].value_counts(dropna=False)

30.0    369323
7.0     320608
6.0     240013
4.0     221696
3.0     217005
5.0     214503
NaN     206209
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
11.0     80970
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: days_since_prior_order, dtype: int64

In [6]:
# Renaming columns
df_ords_2.rename(columns={'order_dow':'order_day_of_week'},inplace=True)

In [7]:
df_ords_2.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [8]:
# Changing data type for necessary columns
df_ords_2['order_number']=df_ords['order_number'].astype('str')

In [9]:
df_ords_2['order_number'].dtype

dtype('O')

In [10]:
df_ords_2['order_id']=df_ords['order_id'].astype('str')

In [11]:
df_ords_2['user_id']=df_ords['user_id'].astype('str')

In [12]:
df_ords_2.dtypes

order_id                   object
user_id                    object
order_number               object
order_day_of_week           int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

In [13]:
df_ords_2.describe()

order_day_of_week  order_hour_of_day  days_since_prior_order
count       3.421083e+06       3.421083e+06            3.214874e+06
mean        2.776219e+00       1.345202e+01            1.111484e+01
std         2.046829e+00       4.226088e+00            9.206737e+00
min         0.000000e+00       0.000000e+00            0.000000e+00
25%         1.000000e+00       1.000000e+01            4.000000e+00
50%         3.000000e+00       1.300000e+01            7.000000e+00
75%         5.000000e+00       1.600000e+01            1.500000e+01
max         6.000000e+00       2.300000e+01            3.000000e+01

In [14]:
# Transposing Data
df_dep.head()

department_id       1      2       3        4        5              6  \
0    department  frozen  other  bakery  produce  alcohol  international   

           7     8                9  ...            12      13         14  \
0  beverages  pets  dry goods pasta  ...  meat seafood  pantry  breakfast   

             15          16         17      18      19    20       21  
0  canned goods  dairy eggs  household  babies  snacks  deli  missing  

[1 rows x 22 columns]

In [15]:
# Transposing df_dep
df_dep_t=df_dep.T
df_dep_t

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [16]:
# adding index to the dataframe
df_dep_t.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

In [17]:
#1a) Create new header to the dataframe
new_header=df_dep_t.iloc[0]
new_header

0    department
Name: department_id, dtype: object

In [18]:
#1b) Remove first row in the header
df_dep_t_new=df_dep_t[1:]
df_dep_t_new

0
1            frozen
2             other
3            bakery
4           produce
5           alcohol
6     international
7         beverages
8              pets
9   dry goods pasta
10             bulk
11    personal care
12     meat seafood
13           pantry
14        breakfast
15     canned goods
16       dairy eggs
17        household
18           babies
19           snacks
20             deli
21          missing

In [19]:
#1c) Add new header
df_dep_t_new.columns=new_header
df_dep_t_new

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [20]:
df_prods

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49693 rows x 5 columns]

In [21]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [22]:
data_dict=df_dep_t_new.to_dict('index')
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

# Task 4.4

In [23]:
# 1) Performed the wrangling procedures 

In [24]:
# 2) All the variable have been changed to suitable format in the data wrangling procedure earlier.[in(33)]

In [25]:
df_ords_2

order_id user_id order_number  order_day_of_week  order_hour_of_day  \
0        2539329       1            1                  2                  8   
1        2398795       1            2                  3                  7   
2         473747       1            3                  3                 12   
3        2254736       1            4                  4                  7   
4         431534       1            5                  4                 15   
...          ...     ...          ...                ...                ...   
3421078  2266710  206209           10                  5                 18   
3421079  1854736  206209           11                  4                 10   
3421080   626363  206209           12                  1                 12   
3421081  2977660  206209           13                  1                 12   
3421082   272231  206209           14                  6                 14   

         days_since_prior_order  
0                           NaN  
1                          15.0  
2                          21.0  
3                          29.0  
4                          28.0  
...                         ...  
3421078                    29.0  
3421079                    30.0  
3421080                    18.0  
3421081                     7.0  
3421082                    30.0  

[3421083 rows x 6 columns]

In [26]:
# 3) changing the name without overwriting in dataframe
df_ords_2.rename(columns={'user_id':'customer_id'}, inplace=True)
df_ords_2

order_id customer_id order_number  order_day_of_week  \
0        2539329           1            1                  2   
1        2398795           1            2                  3   
2         473747           1            3                  3   
3        2254736           1            4                  4   
4         431534           1            5                  4   
...          ...         ...          ...                ...   
3421078  2266710      206209           10                  5   
3421079  1854736      206209           11                  4   
3421080   626363      206209           12                  1   
3421081  2977660      206209           13                  1   
3421082   272231      206209           14                  6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 6 columns]

In [27]:
# 4) Busiest hour for placing orders
df_ords_2['order_hour_of_day'].describe()

count    3.421083e+06
mean     1.345202e+01
std      4.226088e+00
min      0.000000e+00
25%      1.000000e+01
50%      1.300000e+01
75%      1.600000e+01
max      2.300000e+01
Name: order_hour_of_day, dtype: float64

In [28]:
# coverting float to integer
df_ords_2['order_hour_of_day']=df_ords_2['order_hour_of_day'].astype('int')

In [29]:
df_ords_2['order_hour_of_day'].value_counts()

10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: order_hour_of_day, dtype: int64

4) The busiest hour is 10am with 288418 placed orders

In [31]:
# 5) the meaning behind value of 4 in the 'department_id' column
print(data_dict.get('4'))

{'department': 'produce'}


In [32]:
# 6) breakfast item sales
df_dep_t_new
#14 is the id for breakfast 

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [37]:
df_breakfast=df_prods[df_prods['department_id']==14]
df_breakfast

product_id                                       product_name  \
27             28                                  Wheat Chex Cereal   
33             34                                                NaN   
67             68                            Pancake Mix, Buttermilk   
89             90                                       Smorz Cereal   
210           211   Gluten Free Organic Cereal Coconut Maple Vanilla   
...           ...                                                ...   
49330       49326                            Cereal Variety Fun Pack   
49395       49391            Light and Fluffy Buttermilk Pancake Mix   
49547       49543                          Chocolate Cheerios Cereal   
49637       49633               Shake 'N Pour Buttermilk Pancake Mix   
49667       49663  Ultra Protein Power Crunch Peanut Butter N' Ho...   

       aisle_id  department_id  prices  
27          121             14    10.1  
33          121             14    12.2  
67          130             14    13.7  
89          121             14     3.9  
210         130             14     3.6  
...         ...            ...     ...  
49330       121             14     9.1  
49395       130             14     2.0  
49547       121             14    10.8  
49637       130             14    14.2  
49667        57             14    10.2  

[1116 rows x 5 columns]

In [39]:
# 7) Customers purchasing dinner items
df_dinner_items=df_prods.loc[df_prods['department_id'].isin([5,20,7,12])]
df_dinner_items

product_id                                    product_name  aisle_id  \
2               3            Robust Golden Unsweetened Oolong Tea        94   
6               7                  Pure Coconut Water With Orange        98   
9              10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10             11                               Peach Mango Juice        31   
16             17                               Rendered Duck Fat        35   
...           ...                                             ...       ...   
49676       49672                          Cafe Mocha K-Cup Packs        26   
49679       49675             Cinnamon Dolce Keurig Brewed K Cups        26   
49680       49676                          Ultra Red Energy Drink        64   
49686       49682                              California Limeade        98   
49688       49684       Vodka, Triple Distilled, Twist of Vanilla       124   

       department_id  prices  
2                  7     4.5  
6                  7     4.4  
9                  7     8.4  
10                 7     2.8  
16                12    17.1  
...              ...     ...  
49676              7     6.5  
49679              7    14.0  
49680              7    14.5  
49686              7     4.3  
49688              5     5.3  

[7650 rows x 5 columns]

In [40]:
# 8) The last dataframe that i created have 7650 rows

In [45]:
# 9)Extracting info about customer_id: 1
df_customer_id_1=df_ords_2[df_ords_2['customer_id']=='1'] #using quotes for id as we have converted it to string
df_customer_id_1

order_id customer_id order_number  order_day_of_week  order_hour_of_day  \
0   2539329           1            1                  2                  8   
1   2398795           1            2                  3                  7   
2    473747           1            3                  3                 12   
3   2254736           1            4                  4                  7   
4    431534           1            5                  4                 15   
5   3367565           1            6                  2                  7   
6    550135           1            7                  1                  9   
7   3108588           1            8                  1                 14   
8   2295261           1            9                  1                 16   
9   2550362           1           10                  4                  8   
10  1187899           1           11                  4                  8   

    days_since_prior_order  
0                      NaN  
1                     15.0  
2                     21.0  
3                     29.0  
4                     28.0  
5                     19.0  
6                     20.0  
7                     14.0  
8                      0.0  
9                     30.0  
10                    14.0

In [46]:
# 10) Details about customer's behaviour
df_customer_id_1.describe()

order_day_of_week  order_hour_of_day  days_since_prior_order
count          11.000000          11.000000               10.000000
mean            2.636364          10.090909               19.000000
std             1.286291           3.477198                9.030811
min             1.000000           7.000000                0.000000
25%             1.500000           7.500000               14.250000
50%             3.000000           8.000000               19.500000
75%             4.000000          13.000000               26.250000
max             4.000000          16.000000               30.000000

In [62]:
# 12) Export df_ords to prepared data
path=r'C:\Users\ponni\Achievement 4'
df_ords_2.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_wrangled.csv'))

In [63]:
#12 Exporting df_dep_t_new to prepared data
df_dep_t_new.to_csv(os.path.join(path, '02 Data','Prepared Data', 'departments_wrangled.csv'))